### PR TITLE
Add support for passing types and directives when building a schema

### LINKED ISSUE #9: Expose functionality for passing custom directives
buildSchema does seem to support it while, build doesn't

```js
export default function buildSchema(schemaAST, { types, directives } = {}, typeDeps) {
...
}
```

### PR DESCRIPTION
**Pull request by nripendra**
As suggested here: 
https://github.com/bloveit/graphql-utilities/issues/9#issuecomment-262299987

closes bloveit/graphql-utilities#9

### CODE REVIEW

**Comment by migueloller** [comment]
@nripendra, thanks for the PR! Will go over it soon and review.

File: src/build/__tests__/build-test.js [hunk 1]
```diff
@@ -125,8 +144,8 @@ describe('build()', () => {
 
   it('should throw if building a schema and types is not an array', () => {
     const msg = 'Can\'t use thunks as type dependencies for schema.';
-    expect(() => build(schemaSource + querySource, undefined, () => {})).toThrowError(msg);
-    expect(() => build(querySource, undefined, () => {})).toThrowError(msg);
+    expect(() => build(schemaSource + querySource, undefined, () => { })).toThrowError(msg);
```
Comment (migueloller) [inline src/build/__tests__/build-test.js]: Looks like the editor added a space between the brackets. Can we get it to what it was before?

File: src/build/__tests__/build-test.js [hunk 2]
```diff
@@ -125,8 +144,8 @@ describe('build()', () => {
...
+    expect(() => build(querySource, undefined, () => { })).toThrowError(msg);
```
Comment (migueloller) [inline src/build/__tests__/build-test.js]: Same comment as above on bracket space.

File: src/build/index.js [hunk 3]
```diff
@@ -91,9 +91,17 @@ export default function build(source, config = {}, typeDeps = [], infer = true)
     if (!Array.isArray(typeDeps)) {
       throw new Error('Can\'t use thunks as type dependencies for schema.');
     }
+
+    const schemaConfigKey = '__schema';
```
Comment (migueloller) [inline src/build/index.js]: nit: Can we have this constant be declared in the outermost scope with the name `SCHEMA_CONFIG_KEY`?

File: src/build/__tests__/build-test.js (hunk 1 above)
Comment (migueloller) [inline src/build/__tests__/build-test.js]: Looks like the editor added a space between the brackets. Can we get it to what it was before?

File: src/build/__tests__/build-test.js (hunk 2 above)
Comment (migueloller) [inline src/build/__tests__/build-test.js]: Same comment as above on bracket space.

File: src/build/index.js [hunk 4]
```diff
@@ -115,5 +123,5 @@ export default function build(source, config = {}, typeDeps = [], infer = true)
   }
 
   // return a type map
-  return types.reduce((map, type) => ({ ...map, [type.name]: type }), {});
+  return types.reduce((map, type) => ({ ...map, [type.name]: type }), { });
```
Comment (migueloller) [inline src/build/index.js]: Same comment as above on bracket space.

Comment (migueloller) [inline src/build/index.js]: It seems to me that what lines 96 - 106 are doing could be changed to:

```js
return buildSchema(
  schemaASTs[0],
  config[SCHEMA_CONFIG_KEY],
  () => ensureNoDuplicateTypes([...types, ...typeDeps]),
);
```

The union of the type dependencies and schema types into `schemaConfig.types` could cause unexpected side effects as the schema's types and the type dependencies for the template string are 2 different things.

File: src/build/index.js (hunk 3 above)
Comment (migueloller) [inline src/build/index.js]: nit: Can we have this constant be declared in the outermost scope with the name `SCHEMA_CONFIG_KEY`?

File: src/build/__tests__/build-test.js [hunk 5]
```diff
@@ -9,15 +9,30 @@ import {
   GraphQLBoolean,
   GraphQLID,
 } from 'graphql/type';
+import { DirectiveLocation, GraphQLDirective } from 'graphql/type/directives';
```
Comment (migueloller) [inline src/build/__tests__/build-test.js]: nit: These 2 imports are available from `graphql/type` so you could just add them at the end of the import statement above.

File: src/build/__tests__/build-test.js [hunk 6]
```diff
@@ -9,15 +9,30 @@ import {
...
+});
 const querySource = `
  type Query {
    ok: Boolean!
  }
 `;
+const querySourceWithDirective = `
+ type Query {
+   ok: Boolean! @Dir1
+ }
```
Comment (migueloller) [inline src/build/__tests__/build-test.js]: It was my intention to have 2 spaces of indentation for the template string as in `timestampSource` for example. Could you update that for `querySourceWithDirective` and `querySource`? 😄

File: src/build/__tests__/build-test.js [hunk 7]
```diff
@@ -9,15 +9,30 @@ import {
...
 import build, { buildTypes } from '../';
 
+const customDirective = new GraphQLDirective({
+  name: 'Dir1',
+  description:
+  'My custom uppercase',
+  locations: [
+    DirectiveLocation.FIELD,
+  ],
+});
```
Comment (migueloller) [inline src/build/__tests__/build-test.js:21]: A couple of things here (very nitpicky 😅 ).

- For consistency, can we rename the variable `CustomDirective` (pascal case) and change the name field to `'CustomDirective'`?
- For style consistency, can the `description` field value be on the same line instead of a new line?
- Again for consistency, can lines 22 - 24 be condensed into a single line?
- Finally, can the directive description be a little be more descriptive? (e.g., 'Custom directive') You could also simply remove the entire description (probably prefer this as it's unnecessary for the test).

File: src/build/__tests__/build-test.js [hunk 8]
```diff
@@ -9,15 +9,30 @@ import {
...
 const querySource = `
  type Query {
    ok: Boolean!
  }
 `;
+const querySourceWithDirective = `
```
Comment (migueloller) [inline src/build/__tests__/build-test.js]: nit: Again, super nit picky (😅) but can you change the name to `queryWithDirectiveSource`?

File: src/build/__tests__/build-test.js [hunk 9]
```diff
@@ -149,4 +168,16 @@ describe('build()', () => {
       Timestamp: { serialize },
     }, undefined, false)).forEach((type, i) => expectTypesEqual(type, [Record, Timestamp][i]));
   });
+
+  it('should allow directives configuration using __schema', () => {
+    const config = { __schema: { directives: [customDirective] } };
+    const target = build(schemaSource + querySourceWithDirective, config, undefined, false);
```
Comment (migueloller) [inline src/build/__tests__/build-test.js]: You don't need to pass `false` as the last argument because the schema won't be inferred. Once you get rid of the `false` you can also get rid of the `undefined`.

File: src/build/__tests__/build-test.js [hunk 10]
```diff
@@ -149,4 +168,16 @@ describe('build()', () => {
...
+    expectSchemasEqual(target, SchemaWithDirective);
+  });
+
+  it('should not build schema with default directives', () => {
+    const config = { __schema: { directives: [customDirective] } };
+    const target = build(schemaSource + querySourceWithDirective, config, undefined, false);
```
Comment (migueloller) [inline src/build/__tests__/build-test.js]: Same comment as above regarding schema inference.

File: src/build/__tests__/build-test.js [hunk 11]
```diff
@@ -149,4 +168,16 @@ describe('build()', () => {
...
     }, undefined, false)).forEach((type, i) => expectTypesEqual(type, [Record, Timestamp][i]));
   });
+
+  it('should allow directives configuration using __schema', () => {
+    const config = { __schema: { directives: [customDirective] } };
+    const target = build(schemaSource + querySourceWithDirective, config, undefined, false);
+    expectSchemasEqual(target, SchemaWithDirective);
+  });
+
+  it('should not build schema with default directives', () => {
```
Comment (migueloller) [inline src/build/__tests__/build-test.js]: I'm not entirely sure what this test case is testing for. Would you mind explaining what it's doing that the other test case doesn't catch? 😃

**Review by migueloller — CHANGES_REQUESTED** [review]
@nripendra,

Really awesome work with the PR! 🎉 

I made some comments and most of them were nits just to keep the code as clean and consistent to our style as possible.

**Comment by migueloller** [comment]
Another observation is that this addition doesn't work for inferred schemas. This is OK because inferred schemas will be removed in `0.1.0` but it should be clear in the commit messages and elsewhere that the feature is only for defined schemas.

File: src/build/__tests__/build-test.js (hunk 11 above)
Comment (nripendra) [reply src/build/__tests__/build-test.js]: - **'should allow directives configuration using __schema'** tests for to see that we can pass __schema.
- **'should not build schema with default directives'** tests that the schema returned by build is not same when config __schema is passed. This extra test was added because I had one mistake in my setup code but the first test (i.e: 'should allow directives configuration using __schema') didn't catch that mistake.

File: src/build/__tests__/build-test.js (hunk 9 above)
Comment (nripendra) [reply src/build/__tests__/build-test.js]: Didn't get this point... Should I remove the last two arguments (`undefined` and `false`)? Or just `false' only?

Comment (migueloller) [reply src/build/__tests__/build-test.js]: You can remove both.

File: src/build/__tests__/build-test.js (hunk 11 above)
Comment (migueloller) [reply src/build/__tests__/build-test.js]: Hmm, it seems to me that the second test should be in `buildSchema-test.js`. And it would seem that the second test is the base case that's already tested in that file. The only new change in functionality is that `build` should know to pass the object in `__schema` to `buildSchema`.

**Comment by nripendra** [comment]
@migueloller 
I have implemented most of the requested changes, could you please check. Let me know if you think more changes are necessary.

File: src/build/__tests__/build-test.js [hunk 12]
```diff
@@ -8,15 +8,27 @@ import {
   GraphQLNonNull,
   GraphQLBoolean,
   GraphQLID,
+  DirectiveLocation,
+  GraphQLDirective,
 } from 'graphql/type';
+
```
Comment (migueloller) [inline src/build/__tests__/build-test.js]: nit: Can you remove this extra line?

File: src/build/__tests__/build-test.js [hunk 13]
```diff
@@ -149,4 +165,16 @@ describe('build()', () => {
       Timestamp: { serialize },
     }, undefined, false)).forEach((type, i) => expectTypesEqual(type, [Record, Timestamp][i]));
   });
+
+  it('should allow directives configuration using __schema', () => {
+    const config = { __schema: { directives: [CustomDirective] } };
+    const target = build(schemaSource + queryWithDirectiveSource, config);
+    expectSchemasEqual(target, SchemaWithDirective);
+  });
```
Comment (migueloller) [inline src/build/__tests__/build-test.js:173]: Could you import `SCHEMA_CONFIG_KEY` from the main file here so that if the key changes in the future there is no need to change the test?

Would look something like:

```js
it('should allow a schema configuration to be passed via SCHEMA_CONFIG_KEY', () => {
```

This enhancement you're doing "enables" directive support but more generally it allows to configure schema via the `__schema` key, which is what we should test in `build`. If you look at the [tests](https://github.com/bloveit/graphql-utilities/blob/master/src/build/__tests__/buildSchema-test.js#L71) in `buildSchema` you'll see that directive support was already being supported in a way.

File: src/build/__tests__/build-test.js [hunk 14]
```diff
@@ -149,4 +165,16 @@ describe('build()', () => {
...
+
+  it('should not build schema with default directives', () => {
+    const config = { __schema: { directives: [CustomDirective] } };
+    const target = build(schemaSource + queryWithDirectiveSource, config);
+    expect(() => expectSchemasEqual(target, Schema)).toThrowError();
+  });
```
Comment (migueloller) [inline src/build/__tests__/build-test.js]: Like I mentioned in my previous comment, this test case should already be covered in `buildSchema-test.js`, and if it's not then it should be moved there since `build` is only concerned with calling `buildSchema` when necessary and appropriately. The internal works of `buildSchema` should be tested in `buildSchema-test.js`.

I'm still not sure if this test case is needed. It basically checks that when it's built with the directive it doesn't equal a schema that isn't built with it. This is tested by the fact that the first test passes.

If you're trying to check for default directives (e.g., `skip` and `include`) then the test case should look differently.

Could you maybe explain in more detail?

Please check the `buildSchema` [test file](https://github.com/bloveit/graphql-utilities/blob/master/src/build/__tests__/buildSchema-test.js) and you'll see what I mean.

Basically, the only test case needed is the first one you added but it should test if the right config object is being passed to `buildSchema` (which it's already doing but the test description needs to be updated).

File: src/build/index.js [hunk 15]
```diff
@@ -91,9 +93,10 @@ export default function build(source, config = {}, typeDeps = [], infer = true)
     if (!Array.isArray(typeDeps)) {
       throw new Error('Can\'t use thunks as type dependencies for schema.');
     }
+
```
Comment (migueloller) [inline src/build/index.js]: nit: Can you remove this extra line?

File: src/build/index.js [hunk 16]
```diff
@@ -18,6 +18,8 @@ const buildMap = {
   [Kind.INPUT_OBJECT_TYPE_DEFINITION]: buildInputObject,
 };
 
+const SCHEMA_CONFIG_KEY = '__schema';
```
Comment (migueloller) [inline src/build/index.js]: You'll need to export this to be able to test the right configuration is being passed.

**Review by migueloller — CHANGES_REQUESTED** [review]
@nripendra,

Sorry for the delay in responding, the holidays have kept me busy.

Again, thanks for all the hard work in putting up this PR.

Just a few more changes and we'll be ready to merge.

- The first test needs to be made more generic and it's description updated to reflect what is actually being tested.
- The second test is probably unnecessary. Take a look at my comments and the `buildSchema` test file.
- There are some nits regarding some newlines that got inserted.

That's it! 😄 

Let me know if there's anything I can help with.

**Comment by nripendra** [comment]
@migueloller 
Please check now 😃

File: src/build/__tests__/build-test.js [hunk 17]
```diff
@@ -149,4 +164,11 @@ describe('build()', () => {
       Timestamp: { serialize },
     }, undefined, false)).forEach((type, i) => expectTypesEqual(type, [Record, Timestamp][i]));
   });
+
+  it('should allow directives configuration using SCHEMA_CONFIG_KEY', () => {
```
Comment (migueloller) [inline src/build/__tests__/build-test.js]: Please change `directives` to `schema` in the test description and I'll merge! 😄 

The reason for this change is because this change/test actually works for any schema configuration and not just directives.

**Review by migueloller — CHANGES_REQUESTED** [review]
One last small change and I'll merge!

Thanks for all the hard work @nripendra!

File: src/build/__tests__/build-test.js (hunk 17 above)
Comment (nripendra) [reply src/build/__tests__/build-test.js]: Done, please check.

**Review by migueloller — APPROVED** [review]
